impl/src/main/java/io/kroxylicious/testing/kafka/invm/InVMKafkaCluster.java

```diff
@@ -245,12 +244,75 @@ public Map<String, Object> getKafkaClientConfiguration(String user, String passw
+
+        // preallocate ports for the new broker
+        try (var preallocator = new ListeningSocketPreallocator()) {
+            ports.get(Listener.EXTERNAL).put(newNodeId, preallocator.preAllocateListeningSockets(1).get(0));
+            ports.get(Listener.ANON).put(newNodeId, preallocator.preAllocateListeningSockets(1).get(0));
+            ports.get(Listener.INTERNAL).put(newNodeId, preallocator.preAllocateListeningSockets(1).get(0));
+        }
+
+        var configHolder = clusterConfig.generateConfigForSpecificBroker(this, newNodeId);
+        final Server server = buildKafkaServer(configHolder);
+        tryToStartServerWithRetry(configHolder, server);
+        servers.put(configHolder.getBrokerNum(), server);
+
+        Utils.awaitExpectedBrokerCountInClusterViaTopic(
+                clusterConfig.getConnectConfigForCluster(buildServerList(brokerId -> getEndpointPair(Listener.ANON, brokerId)), null, null, null, null), 120,
+                TimeUnit.SECONDS,
+                getNumOfBrokers());
+        return configHolder.getBrokerNum();
+    }
+
+    @Override
+    public synchronized void removeBroker(int nodeId) throws IllegalArgumentException, UnsupportedOperationException {
+        if (!servers.containsKey(nodeId)) {
+            throw new IllegalArgumentException("Broker node " + nodeId + " is not a member of the cluster.");
+        }
+        if (clusterConfig.isKraftMode() && ports.get(Listener.CONTROLLER).containsKey(nodeId)) {
+            throw new UnsupportedOperationException("Cannot remove controller node " + nodeId + " from a kraft cluster.");
+        }
+        if (servers.size() < 2) {
+            throw new IllegalArgumentException("Cannot remove a node from a cluster with only %d nodes".formatted(servers.size()));
+        }
+
+        var target = servers.keySet().stream().filter(n -> n != nodeId).findFirst();
+        if (target.isEmpty()) {
+            throw new IllegalStateException("Could not identify a node to be the re-assignment target");
+        }
+
+        Utils.awaitReassignmentOfKafkaInternalTopicsIfNecessary(
+                clusterConfig.getConnectConfigForCluster(buildServerList(brokerId -> getEndpointPair(Listener.ANON, brokerId)), null, null, null, null), nodeId,
+                target.get(), 120, TimeUnit.SECONDS);
+
+        ports.values().forEach(pm -> pm.remove(nodeId));
+
+        var s = servers.remove(nodeId);
+        s.shutdown();
+        s.awaitShutdown();
+    }
+
     @SuppressWarnings("ResultOfMethodCallIgnored")
     @Override
-    public void close() throws Exception {
+    public synchronized void close() throws Exception {
         try {
             try {
-                servers.stream().parallel().forEach(Server::shutdown);
+                // shutdown controller last seems to be significant to the kraft case.
+                servers.entrySet().stream()
+                        .sorted(Map.Entry.comparingByKey(Comparator.reverseOrder()))
```

Comment:
I'm guessing this works because the controller is just the first broker started (assuming we use combined mode for the broker & controller)? Should we be doing something to more explicitly track which node(s) are controllers and make sure they are shut down after?

Reply:
> I'm guessing this works because the controller is just the first broker started (assuming we use combined mode for the broker & controller)?

that's it.

It feels like we are hitting a kraft bug.  It seems `parallel()` was actually masking it.  `parallel()` is a no-op for streams derived from a treemap so that is how it was uncovered.  I'll see if I can put together a kafka only reproducer (separate from this work).

> Should we be doing something to more explicitly track which node(s) are controllers and make sure they are shut down after?

We have the information we need to do this already. I'll make a change.